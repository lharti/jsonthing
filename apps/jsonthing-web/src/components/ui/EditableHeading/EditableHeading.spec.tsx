import { act, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import { EditableHeading, UIMode } from './index'

describe('<EditableHeading />', () => {
    it('should render with view ui mode by default', () => {
        expect.assertions(2)

        const { container } = render(
            <EditableHeading initialValue="TEXT_VALUE" />,
        )

        const heading = screen.getByRole('heading')

        expect(container).toMatchSnapshot('EditableHeading - View mode')

        expect(heading).not.toHaveFocus()
    })

    it('should render edit ui mode', () => {
        expect.assertions(2)

        const { container } = render(
            <EditableHeading
                initialValue="TEXT_VALUE"
                initialUIMode={UIMode.EDIT}
            />,
        )

        const heading = screen.queryByRole('input')

        expect(container).toMatchSnapshot('EditableHeading - Edit mode')

        expect(heading).toHaveFocus()
    })

    describe('heading', () => {
        it('should switch to edit mode on focus', () => {
            expect.assertions(1)

            render(<EditableHeading initialValue="TEXT_VALUE" />)

            const heading = screen.getByRole('heading')

            act(() => {
                heading.focus()
            })

            const headingInput = screen.queryByRole('input')

            expect(headingInput).toBeInTheDocument()
        })

        it('should switch to view mode on Enter key press', async () => {
            expect.assertions(2)

            const user = userEvent.setup()

            render(
                <EditableHeading
                    initialValue="INITIAL_VALUE"
                    initialUIMode={UIMode.EDIT}
                />,
            )

            const heading = screen.getByRole('input')

            await act(async () => {
                await user.type(heading, '{enter}')
            })

            expect(heading).toHaveAttribute('role', 'heading')

            expect(heading).not.toHaveFocus()
        })

        it('should call onChange with new value on enter key press', async () => {
            expect.assertions(1)

            const user = userEvent.setup()

            const initialValue = 'INITIAL_VALUE'
            const onChange = jest.fn()

            render(
                <EditableHeading
                    initialValue={initialValue}
                    onChange={onChange}
                />,
            )

            const heading = screen.getByRole('heading')

            const randomText = Math.random().toString()

            await act(async () => {
                await user.click(heading)
                await user.type(heading, `${randomText}{enter}`)
            })

            const newHeadingText = initialValue + randomText

            expect(onChange).toHaveBeenCalledExactlyOnceWith(newHeadingText)
        })

        it('should update textContent with typed values', async () => {
            expect.assertions(1)

            const user = userEvent.setup()

            const initialValue = 'INITIAL_VALUE'

            render(
                <EditableHeading
                    initialValue={initialValue}
                    initialUIMode={UIMode.EDIT}
                />,
            )

            const heading = screen.getByRole('input')

            const randomText = Math.random().toString()

            await act(async () => {
                await user.type(heading, `${randomText}`)
            })

            const newHeadingText = initialValue + randomText

            expect(heading).toHaveTextContent(newHeadingText)
        })
    })

    describe('edit button', () => {
        it('should switch to edit mode on click', () => {
            expect.assertions(1)

            render(<EditableHeading initialValue="TEXT_VALUE" />)

            const editButton = screen.getByRole('button')

            act(() => {
                editButton.click()
            })

            const headingInput = screen.queryByRole('input')

            expect(headingInput).toBeInTheDocument()
        })

        it('should switch to view mode on click if current mode is edit', () => {
            expect.assertions(1)

            render(
                <EditableHeading
                    initialValue="TEXT_VALUE"
                    initialUIMode={UIMode.EDIT}
                />,
            )

            const editButton = screen.getByRole('button')

            act(() => {
                editButton.click()
            })

            const heading = screen.queryByRole('heading')

            expect(heading).toBeInTheDocument()
        })

        it('should call onChange with new value on click if current mode is edit', async () => {
            expect.assertions(1)

            const user = userEvent.setup()

            const initialValue = 'INITIAL_VALUE'
            const onChange = jest.fn()

            render(
                <EditableHeading
                    initialValue={initialValue}
                    initialUIMode={UIMode.EDIT}
                    onChange={onChange}
                />,
            )

            const editButton = screen.getByRole('button')
            const headingInput = screen.getByRole('input')

            const randomText = Math.random().toString()

            await act(async () => {
                await user.type(headingInput, `${randomText}`)

                await user.click(editButton)
            })

            const newHeadingText = initialValue + randomText

            expect(onChange).toHaveBeenCalledExactlyOnceWith(newHeadingText)
        })
    })
})
