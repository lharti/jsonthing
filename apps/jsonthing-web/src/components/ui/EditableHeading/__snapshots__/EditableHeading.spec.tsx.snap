// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditableHeading /> should render edit ui mode: EditableHeading - Edit mode 1`] = `
<div>
  <h1
    class="
                  cursor-pointer text-3xl font-bold outline outline-2
                  outline-transparent

                  focus:cursor-auto focus:rounded-sm focus:outline-blue-300
                "
    contenteditable="true"
    role="input"
    spellcheck="false"
  >
    TEXT_VALUE
  </h1>
  <button
    aria-label="edit title"
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors dark:focus-visible:ring-gray-300 disabled:pointer-events-none disabled:opacity-50 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-950 border border-gray-200 bg-white shadow-sm dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 hover:bg-gray-100 hover:text-gray-900 h-9 ml-2 p-1"
    role="button"
  >
    <svg
      class="tabler-icon tabler-icon-check "
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5 12l5 5l10 -10"
      />
    </svg>
  </button>
</div>
`;

exports[`<EditableHeading /> should render with view ui mode by default: EditableHeading - View mode 1`] = `
<div>
  <h1
    class="
                  cursor-pointer text-3xl font-bold outline outline-2
                  outline-transparent

                  focus:cursor-auto focus:rounded-sm focus:outline-blue-300
                "
    contenteditable="true"
    role="heading"
    spellcheck="false"
  >
    TEXT_VALUE
  </h1>
  <button
    aria-label="edit title"
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors dark:focus-visible:ring-gray-300 disabled:pointer-events-none disabled:opacity-50 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 hover:bg-gray-100 hover:text-gray-900 h-9 ml-2 p-1"
    role="button"
  >
    <svg
      class="tabler-icon tabler-icon-pencil "
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4"
      />
      <path
        d="M13.5 6.5l4 4"
      />
    </svg>
  </button>
</div>
`;
